File: pyps/src/main/scala/com/tencent/angel/ps/variable/VariableMeta.java
package com.tencent.angel.ps.variable;

import com.tencent.angel.common.Meta;
import com.tencent.angel.matrix.MatrixContext;
import com.tencent.angel.model.MatrixLoadContext;
import com.tencent.angel.model.MatrixSaveContext;


import java.util.ArrayList;

public class VariableMeta extends Meta {
    private int numSlot;

    public VariableMeta(String name, String dtype, int dim, long[] shape, long validIndexNum, int numSlot) {
        super(name, dtype, dim, shape, validIndexNum);
        assert numSlot >= 1;
        this.numSlot = numSlot;

        assert dim == shape.length;
        if (rowType.isDense()) {
            int numRows = numSlot + 1;
            long numCols = 1L;
            for (int i = 0; i < dim; i++) {
                numCols *= shape[i];
            }

            matrixContext = new MatrixContext(name, numRows, numCols, validIndexNum, -1, -1, rowType);
        } else if (rowType.isSparse()) {
            assert dim <= 2 && dim > 0;
            if (dim == 1) {
                int numRows = numSlot + 1;
                long numCols = shape[0];
                matrixContext = new MatrixContext(name, numRows, numCols, validIndexNum, -1, -1, rowType);
            } else {
                int numRows = (numSlot + 1) * (int) shape[0];
                long numCols = shape[1];
                matrixContext = new MatrixContext(name, numRows, numCols, validIndexNum, -1, -1, rowType);
            }
        }

    }

    public VariableMeta(String name, String dtype, int dim, long[] shape, int numSlot) {
        this(name, dtype, dim, shape, -1, numSlot);
    }

    public VariableMeta(String name, String dtype, int dim, long[] shape) {
        this(name, dtype, dim, shape, -1, 1);
    }

    public VariableMeta(String name, String dtype, int numRows, long numCols) {
        this(name, dtype, numRows, numCols, -1);
    }

    public VariableMeta(String name, String dtype, int numRows, long numCols, long validIndexNum) {
        this(name, dtype, 2, new long[]{numRows, numCols}, validIndexNum, 1);
    }

    public MatrixContext getMatrixContext() {
        return matrixContext;
    }

    public MatrixSaveContext getMatrixSaveContext(String path, String formatClassName) {
        ArrayList<Integer> list = new ArrayList<Integer>();
        int originRows = matrixContext.getRowNum() / (numSlot + 1);
        for (int i = 0; i < originRows; i++) {
            list.add(i);
        }

        return new MatrixSaveContext(name, list, formatClassName);
    }

    public MatrixLoadContext getMatrixLoadContext(String path) {
        return new MatrixLoadContext(name, path);
    }

    public int getNumSlot() {
        return numSlot;
    }

    @Override
    public boolean equals(Object obj) {
        if (super.equals(obj)) {
            return this.numSlot == ((VariableMeta) obj).numSlot;
        } else {
            return false;
        }
    }
}

File: pyps/src/main/scala/com/tencent/client/worker/ps/variable/VariableMeta.scala
package com.tencent.client.worker.ps.variable

import java.util

import com.tencent.client.common.Meta
import com.tencent.angel.matrix.MatrixContext
import com.tencent.angel.model.MatrixLoadContext
import com.tencent.angel.model.MatrixSaveContext


class VariableMeta(name: String, dtype: String, dim: Int, shape: Array[Long], validIndexNum: Long, val numSlot: Int)
  extends Meta(name, dtype, dim, shape, validIndexNum) {

  override protected val matrixContext: MatrixContext = if (rowType.isDense) {
    val numRows = numSlot + 1
    val numCols = shape.product

    new MatrixContext(name, numRows, numCols, validIndexNum, -1, -1, rowType)
  } else if (rowType.isSparse) {
    assert(dim <= 2 && dim > 0)
    if (dim == 1) {
      val numRows = numSlot + 1
      val numCols = shape(0)
      new MatrixContext(name, numRows, numCols, validIndexNum, -1, -1, rowType)
    } else {
      val numRows = (numSlot + 1) * shape(0).asInstanceOf[Int]
      val numCols = shape(1)
      new MatrixContext(name, numRows, numCols, validIndexNum, -1, -1, rowType)
    }
  } else {
    throw new Exception("cannot create matrix!")
  }

  def this(name: String, dtype: String, dim: Int, shape: Array[Long], numSlot: Int) {
    this(name, dtype, dim, shape, -1, numSlot)
  }

  def this(name: String, dtype: String, dim: Int, shape: Array[Long]) {
    this(name, dtype, dim, shape, -1, 1)
  }

  def this(name: String, dtype: String, numRows: Int, numCols: Long, validIndexNum: Long) {
    this(name, dtype, 2, Array[Long](numRows, numCols), validIndexNum, 1)
  }

  def this(name: String, dtype: String, numRows: Int, numCols: Long) {
    this(name, dtype, numRows, numCols, -1)
  }

  def getMatrixContext: MatrixContext = matrixContext

  def getMatrixSaveContext(path: String, formatClassName: String): MatrixSaveContext = {
    val list = new util.ArrayList[java.lang.Integer]()
    val originRows = matrixContext.getRowNum / (numSlot + 1)
    (0 until originRows).foreach(i => list.add(i))

    new MatrixSaveContext(name, list, formatClassName)
  }

  def getMatrixLoadContext(path: String): MatrixLoadContext = {
    new MatrixLoadContext(name, path)
  }


  override def equals(obj: Any): Boolean = {
    if (super.equals(obj)) {
      this.numSlot == obj.asInstanceOf[VariableMeta].numSlot
    } else {
      false
    }
  }
}
